lib/api/objectPutPart.js

```diff
@@ -72,95 +36,165 @@ function doStore(objectKeyContext, cipherBundle, request, log,
+    // of at least 5MB until last part.  Also, we check that part sizes are
+    // large enough when mutlipart upload completed.
+    if (Number.parseInt(size, 10) > 5368709120) {
+        return cb(errors.EntityTooLarge);
+    }
+
     const partNumber = Number.parseInt(request.query.partNumber, 10);
     // AWS caps partNumbers at 10,000
     if (partNumber > 10000) {
         return cb(errors.TooManyParts);
     }
+
     if (!Number.isInteger(partNumber) || partNumber < 1) {
         return cb(errors.InvalidArgument);
     }
-    // If part size is greater than 5GB, reject it
-    if (Number.parseInt(size, 10) > 5368709120) {
-        return cb(errors.EntityTooLarge);
-    }
-    // Note: Parts are supposed to be at least 5MB except for last part.
-    // However, there is no way to know whether a part is the last part
-    // since keep taking parts until get a completion request.  But can
-    // expect parts of at least 5MB until last part.  Also, we check that
-    // part sizes are large enough when mutlipart upload completed.
 
-    // Note that keys in the query object retain their case, so
-    // request.query.uploadId must be called with that exact
-    // capitalization
-    const uploadId = request.query.uploadId;
-    const metadataValMPUparams = {
-        authInfo,
-        bucketName,
-        objectKey,
-        uploadId,
-        requestType: 'putPart or complete',
-        log,
-        splitter: constants.splitter,
-    };
-    // For validating the request at the destinationBucket level
-    // params are the same as validating at the MPU level
-    // but the requestType is the more general 'objectPut'
-    const metadataValParams = Object.assign({}, metadataValMPUparams);
-    metadataValParams.requestType = 'objectPut';
+    const bucketName = request.bucketName;
+    assert.strictEqual(typeof bucketName, 'string');
+
+    const canonicalID = authInfo.getCanonicalID();
+    assert.strictEqual(typeof canonicalID, 'string');
+
     log.trace('owner canonicalid to send to data', {
         canonicalID: authInfo.getCanonicalID,
     });
-    const objectKeyContext = {
-        bucketName,
-        owner: authInfo.getCanonicalID(),
-        namespace: request.namespace,
-    };
-
-    return services.metadataValidateAuthorization(
-        metadataValParams,
-        (err, bucket) => {
-            if (err) {
-                return cb(err);
+
+    // Note that keys in the query object retain their case, so
+    // `request.query.uploadId` must be called with that exact capitalization.
+    const uploadId = request.query.uploadId;
+    const mpuBucketName = `${constants.mpuBucketPrefix}${bucketName}`;
+
+    // If bucket has no server-side encryption, `cipherBundle` remains `null`.
+    let cipherBundle = null;
+    let splitter = constants.splitter;
+
+    return async.waterfall([
+        // Get the bucket.
+        next => metadata.getBucket(bucketName, log, next),
```

Comment:
I'd prefer going a little longer form here so we can log here if there is an error getting the bucket.